### ISSUE TITLE
Error: Get list of notifications failed.

### ISSUE THREAD

**Issue by Offerel** [Offerel/SyncMarks-Webapp#62]
Next step was not without problem. When I configure login from Firefox/Syncmark 1.17.9 it cant't save login if I'm not logged in in another tab. When I restart Firefox the log in Firefox/SyncMarks is:
```
2021-03-12 17:22:14 - Info: AddOn version: 1.17.9
2021-03-12 17:22:14 - Info: Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:86.0) Gecko/20100101 Firefox/86.0
2021-03-12 17:22:14 - Info: Current architecture: x86-64 | Current OS: win
2021-03-12 17:22:14 - Info: Initiate PHP startup sync
2021-03-12 17:22:14 - Info: Get list of clients.
2021-03-12 17:22:14 - Info: Get notifications for current client.
2021-03-12 17:22:14 - Info: Initiate startup sync
2021-03-12 17:22:14 - Error: Get list of notifications failed.
2021-03-12 17:22:15 - Info: There was a error retrieving the bookmarks from the server. The status response is: 401
2021-03-12 17:22:15 - Error: Get list of clients failed.:
```
If I try to save from the login screen, I got error. If I then open another tab and login I can save, but not export.
Is there any limit of how many bookmarks I can export? I have more than 2500.
I suppose this could be a new issue....
Regards Johan

_Originally posted by @johmor52 in https://github.com/Offerel/SyncMarks-Webapp/issues/61#issuecomment-797613887_

**Comment by Offerel** [comment]
I have made a new issue for this. This seems to be a server side issue. Can you please add the logfile for the webapp? Per default, you will find this at `/var/log/syncmarks.log`

**Comment by johmor52** [comment]
Hello!
First I saved log with debugmode. Then I emptied logfile and chnged to level 8.
I started Firefox in another computer which has syncmarks-extension. Tried to save - export but no success.
Opened a new tab and logged in to SyncMark-web. Then I could save, but export didn't work.
Save log as syncmarsk-8.log.
Hopefully you can find out the problem
Regards Johan
[syncmarks-8.log](https://github.com/Offerel/SyncMarks-Webapp/files/6132831/syncmarks-8.log)
[syncmarks-9.log](https://github.com/Offerel/SyncMarks-Webapp/files/6132832/syncmarks-9.log)

**Comment by Offerel** [comment]
I have made some tests. At least here it doesn't make a difference, if the WebApp is opened in a tab or not. After installation of the WebApp, i can login with the Plugin by using the password from `config.inc.php` without any problem. If i change the password within the WebApp and restart or reload the Plugin, i get as expected a error with status code 401. When i now change the password in the plugin, to the one changed before in the WebApp, the login is successfully. 

Status 401 means in this case, that the user is not authenticated successfully. I would suggest, you check if your credentials at WebApp and AddOn are identical.

Ah there are the logfiles. As assumed, there must be something wrong with your credentials. There is s switch in the logfile from the user admin to the user johan@hemkonto.se. Since the login with the user johan@hemkonto.se is successful, you should use this user in WebApp and also the Plugin.

```
[12-Mar-2021 17:36:12] - [debug] - admin - 31.208.210.52 - SELECT * FROM `bookmarks` WHERE `bmAction` IS NULL AND `userID` = 1;
[12-Mar-2021 17:53:36] - [notice] -  - 31.208.210.52 - Check login
[12-Mar-2021 17:54:05] - [notice] -  - 31.208.210.52 - Check login
[12-Mar-2021 17:54:05] - [debug] -  - 31.208.210.52 - SELECT * FROM `users` WHERE `userName`= 'johan@hemkonto.se';
[12-Mar-2021 17:54:05] - [notice] - johan@hemkonto.se - 31.208.210.52 - Login successfully
```

**Comment by johmor52** [comment]
I have checked that I have the same credentials at the web and extension.
When I set:
Server to https://hem****/SyncMarks/index.php
Username to johan@hem**** and the correct password
I got login failed: Please check username and password
Then I open a new tab and go to webapp. Login with exactly the same credentials. Thats OK
Then when I save in the other tab I got: Login successfully. Options saved.
- Very strange! It's the same in the other computer.
If I export and say no to delete at web, it doesn't work. But if I say Yes to delete I got a message that it has been uploaded.
I have now changed to adminlevel att my account and then I can see what I have uploaded. I thought that admin could see everything, but I must login with the same as extension to see that users upload.
I still have problem with sync as I got **Error: Get list of clients failed** When I start Firefox.
At least I have uploaded my 2500 links.....

**Comment by Offerel** [comment]
I'm not aware of a limitation of bookmark count. Maybe it takes longer, but there should be no error resulting from a large bookmark count. But let us do this step by step:

- set debug level to 9, to log also SQL requests for the Webapp
- open the webapp in some tab and login
- check that the used credentials are working and you see at least the initial bookmark
- check for error in the logfile. If you are an admin, you can do that from the weapp, if you are no admin you can use a console, like putty and the command "`tail -f /var/log/syncmarks.log`". Let it open to check the log for the next steps
- close the tab with the webapp, if you want
- Open plugin settings
Enter the same credentials as in the webapp.
- click "save"
- in the console, you should something about a testrequest and that login was successful and data is saved, in the client as green toast notification
- click "Export" in the plugin. This can take a while with 2500 bookmarks, you can watch the tail console. After completion, you get a notification in the browser and on console
- open the webapp again, check if the bookmarks are visible

**Comment by johmor52** [comment]
If I use the extension without login in another tab it doesn't work.
I have to have the whole path to webfile, and not www (in my case)
If I export I got export success before it was, but be ready a while later.
If I import I got no message but it works alright. (even with 2500 links)
If I close the other tab with login, it stops working again.
As a result of that it doesn't work to have extension to start at webreader start.
This behavior is for latest Firefox. In Edge I couldn't fill in the credentials. I hvent got the extension for chrome yet (where to find).
Here are my log from web and Firefox:
[Loggfile1.17.9.txt](https://github.com/Offerel/SyncMarks-Webapp/files/6134680/Loggfile1.17.9.txt)
[syncmarks.log](https://github.com/Offerel/SyncMarks-Webapp/files/6134681/syncmarks.log)

The best use of this app would be to have all your bookmarks on the web. Then every reader syncing when starting and transfer/delete both ways all changes to the database on the web.
If you delete one bookmark on the computer, the web database will delete that bookmark the next start of the reader.
Best to have web database as the reference, but with possibillity to change localy and update when starting any webreader.
Regards Johan

**Comment by Offerel** [comment]
Sorry loosed my previous comment. I try to reproduce it again. 
> If I use the extension without login in another tab it doesn't work.

The AddOn uses separate requests and didn't re-use the established session from the WebApp in another tab. However, in some situations, the WebApp can indeed re-use a session from the AddOn. This did not work really reliable in Firefox. It's better in Edge and other Chromium derivatives. Firefox has some open bugs there. But this indicates, that something for some reason wrong data will be send to the Wackend/WebApp.

> I have to have the whole path to webfile, and not www (in my case)

You should use the same url, as with the Backend/WebApp. What url this is, completely depends how you have configured this in your webserver. At least in my case, i can omit www and the filename. I use `https://dev.mydomain.de/bookmarks/`.

> If I export I got export success before it was, but be ready a while later.

Could be the case. This is a asynchronous call. The server can respond with success before it completely finish all SQL queries.

> If I import I got no message but it works alright. (even with 2500 links)

Maybe, i have not configured a notification for the import. I'm not sure. If that's not the case, i will implement this in the next release.

> If I close the other tab with login, it stops working again.

Thats really strange. Are you sure that the AddOn has saved its settings? And that the values like URL, username and password correct and without typos? If there is no Backend/WebApp tab open, i would need to know, the content of the request to analyse it. You can inspect whats saved for your AddOn, when you open `about:devtools-toolbox?id=davmarks%40example.org&type=extension` in Firefox and click on extension storage. wdurl should identical to the Backend/WebApp URL. It must be a https URL. Is a username and password saved and identical to what you want to use?

![grafik](https://user-images.githubusercontent.com/14041522/111032336-ee276d00-840b-11eb-8775-3b25ae6ea751.png)

> As a result of that it doesn't work to have extension to start at webreader start.
> This behavior is for latest Firefox. In Edge I couldn't fill in the credentials. I hvent got the extension for chrome yet (where to find).

Yeah Chrome. I will not upload the extension to the Chrome Webstore. They want me that i should pay a fee, to place and AddOn what i have developed myself there for free. I will not support such behavior. But you can use the *.crx file from the [releases page](https://github.com/Offerel/SyncMarks-Extension/releases) on GitHub and sideload it in Chrome. For Edge its available in their official store, but takes some days longer to review a new version. For Opera it's in analysis since summer last year, waiting for moderation of their fabulous moderation team. But you can use the same *.crx as for Chrome. This works also with Kiwi Browser on Android.

> Here are my log from web and Firefox:
> [Loggfile1.17.9.txt](https://github.com/Offerel/SyncMarks-Webapp/files/6134680/Loggfile1.17.9.txt)
> [syncmarks.log](https://github.com/Offerel/SyncMarks-Webapp/files/6134681/syncmarks.log)

I can't see much new in the logfiles. The problem starts at `2021-03-13 11:30:16 - Error: Get list of notifications failed.` Something seems to prevent communication with the Backend/WebApp. Maybe a wrong URL, wrong credentials or some other thing, like SSL handshake will fail for some reason. Please check the above mentioned `about:devtools-toolbox?id=davmarks%40example.org&type=extension`, to check if URL and credentials are correct. If thats the case, we should check what the browser sends to the backend. You can inspect this, when click in `about:devtools-toolbox?id=davmarks%40example.org&type=extension` on the tab "Console". You can switch frames/tabs by clicking 
![grafik](https://user-images.githubusercontent.com/14041522/111032125-12367e80-840b-11eb-9275-20e29c21f4f9.png). You will then see all the XHR requests, send to your backend.

> The best use of this app would be to have all your bookmarks on the web. Then every reader syncing when starting and transfer/delete both ways all changes to the database on the web.
> If you delete one bookmark on the computer, the web database will delete that bookmark the next start of the reader.
> Best to have web database as the reference, but with possibillity to change localy and update when starting any webreader.
> Regards Johan

Thats why i have created this tool. I cant use Firefox at work and have to use Edge there. But with this, i can open my bookmarks via the backend in every browser. I can add, delete and move bookmarks and they are automatically changed, next time, when im at home.

**Comment by johmor52** [comment]
I am absolutly shure that I use the correct credentials.
I have tried to inspect handshake, but can't see any password. Probably I don't do exactly what you say...

I have tried install addon in chrome, but 1.17.9 give same problem as for Edge: Can't write any credentials. I tried 1.17.3 then I can get it to login, but only if I have another tab open.... Exactly as for Firefox.
If I try to import it'll fail. No import, just deleting the old links in reader.
I copied logfile and from screen.
[Firefox extension.txt](https://github.com/Offerel/SyncMarks-Webapp/files/6135047/Firefox.extension.txt)
[Loggfile1.17.9.txt](https://github.com/Offerel/SyncMarks-Webapp/files/6135049/Loggfile1.17.9.txt)

Do you think it depends on my Swedish version of Firefox? I tried to change to English, but the same result.
I think you have a very good app on its way to be bugfree.
Regards Johan

**Comment by Offerel** [comment]
I think, we come closer to the problem. If i look at Firefox extension.txt, i see errors like:

```
Uncaught ReferenceError: notify is not defined
    onload moz-extension://33ee654b-8ee1-4e0c-b8fd-8c675d7cbb4f/scripts/options.js:110
options.js:110:5
    onload moz-extension://33ee654b-8ee1-4e0c-b8fd-8c675d7cbb4f/scripts/options.js:110
Syncmarks Error: Login failed: Please check username and password. options.js:53:15
    onload moz-extension://33ee654b-8ee1-4e0c-b8fd-8c675d7cbb4f/scripts/options.js:53
Uncaught ReferenceError: notify is not defined
    onload moz-extension://33ee654b-8ee1-4e0c-b8fd-8c675d7cbb4f/scripts/options.js:91
options.js:91:5
    onload moz-extension://33ee654b-8ee1-4e0c-b8fd-8c675d7cbb4f/scripts/options.js:91
Use of nsIFile in content process is deprecated.
```

It didn't find the function `notify()`. As typical for browsers, most code after such an error will not execute. I can solve this error. However this code on line 110 is only triggered, after getting the client name fails for some reason. I have to find, why the request (get client name) fails in your case. To my knowledge it fails, because you get a status 401 from the backend (cant login at the specified url or login fails).

**Comment by Offerel** [comment]
We can make a small test. 

1. If get it right, your Backend is at https://hemkonto.se/SyncMarks/index.php. (are you sure you use the capitals?). 
2. If thats the case, open the URL in a tab, login as admin. 
3. Add a testuser with a simple username like 'johan2@hemkonto.se'. 
4. Add a simple password. Use only alphanumeric signs (letters and ciphers). 
5. Give the user only standard privileges. 
6. Save the user and watch the logfile at the server, if saving was without errors.
7. Logout from WebApp.
8. Login with this fresh created user. 
9. If that works, open AddOn options. 
10. Add https://hemkonto.se/SyncMarks/index.php as URL, johan2@hemkonto.se as the username and the password for this user. 
11. Check if there is no blank at the start or end, when you copy&paste. 
12. Save the options and check the logfile at the server if the testrequest was made and if it was successful. 
13. Close all browser windows when it was successful. 
14. Re-open the browser, go to plugin options, check if data is restored in the fields. 
15. If thats the case, click Export, to export the bookmarks and send them to the server. 
16. Open a tab for the backend (you can click at the plugin name at the options page to do that) 
17. check if all bookmarks are there. 
18. To prevent the notify() error, you can give the client a name in the options page (step 10). But it should work also without a dedicated name. In this case it will use the clientid as name.

**Comment by johmor52** [comment]
I've made the new user. The client name are lost even if I managed to save (had to login as before, this time with admin)
When export I got Error 401 at screen.
I couldn't save if I didn't have a tab or window open to the webapp.
I really have checked the  credentials everytime from the beginning.

**Comment by Offerel** [comment]
Ok, if the mail in your logfile is correct, I will make a test account later on my server and send the credentials to you. If you can successful login there, the problem seems to be your webserver or php installation. Let me know if I can use the mail address

**Comment by johmor52** [comment]
Yes you can use the address.
Regards Johan

**Comment by johmor52** [comment]
Sorry to say it's the same with login only after I have logged in in another tab at your site. I passed the SSL_ERROR_BAD_CERT_DOMAIN, and tested the other domain too.
I also tested with my webapp from Linux-Buster with the same result.
There must be something in the addon not working for me in three different computers.
The addon doesn't work in Edge and Chrome for me. In the older 1.17.3 I could put in the  credentials, but not having the addon to login.
Thank you for tonight!
